src/django_pg_migration_tools/operations.py

```diff
@@ -525,3 +583,256 @@ def describe(self) -> str:
+
+        Reentrancy:
+          Any of the given operations below may fail for a variety of
+          reasons. If a failure occurs, this routine can be rerun without
+          impacting correct execution. Each step progresses toward the
+          outcome while leaving the db in a consistent state.
+          If the operation fails mid-way through, it can still be picked up
+          again. The database isn't left in an inconsistent state.
+
+        Idempotency:
+          The code expects clients to retry migrations without extra
+          side-effects. So does this routine. Multiple calls to the
+          routine have the same effect on the system state as a single
+          call.
+
+        Small number of introspective SQL queries:
+          Introspective SQL queries are necessary for checking the state of
+          the database. This is required for idempotency and reentrancy. At
+          each step, the routine only fires as few introspective SQL
+          statements as necessary.
+        """
+        psql_operations.NotInTransactionMixin()._ensure_not_in_transaction(
+            schema_editor
+        )
+        if not self.allow_migrate_model(schema_editor.connection.alias, model):
+            return
+
+        table_name = model._meta.db_table
+        constraint_name = self._get_constraint_name(table_name, column_name)
+
+        # Just checking for NOT NULL isn't sufficient, we need to check that
+        # the constraint doesn't exist (VALID or NOT). Even a NOT VALID
+        # constraint would set the field to be non-nullable.
+        is_not_null = self._is_not_null(schema_editor, table_name, column_name)
+        constraint_exists = self._constraint_exists(schema_editor, constraint_name)
+        if is_not_null and (not constraint_exists):
+            return
+
+        if not constraint_exists:
+            self._alter_table_not_null_not_valid_constraint(
+                schema_editor, table_name, column_name, constraint_name
+            )
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        elif self._is_constraint_valid(schema_editor, constraint_name):
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
+        else:
+            # Constraint exists and is NOT VALID.
+            self._validate_constraint(schema_editor, table_name, constraint_name)
+            self._alter_table_not_null(schema_editor, table_name, column_name)
+            self._alter_table_drop_constraint(
+                schema_editor, table_name, constraint_name
+            )
+            return
```

Comment:
At the expense of adding a `not` to the `elif` condition, I would swap these two cases. That way the three cases are in order of starting from scratch, picking up the process after 1 step was completed (adding the constraint), or after 2 steps were completed (adding the constraing and validating it).

⚠️  What if the process only completed 3 steps, resulting in the column being NOT NULL but still leaving the constraint behind? It looks like that case isn't dealt with.

Reply:
Thinking about this bit of code some more, I wonder whether it might be clearer (and also reduce code duplication) to structure it as follows:

```python
if need_to_do_step_1:
    do_step1()

if need_to_do_step_2:
    do_step2()

if need_to_do_step_3:
    do_step3()

if need_to_do_step_4:
    do_step4()
```

In normal operation, that would be sequence of four `if True` conditions, and all four steps would be done. But if there had been a failure previously, it would pick up with whichever step needed to be done next.

That might require some extra SQL queries though — I haven't checked.

Reply:
>  leaving the constraint behind? It looks like that case isn't dealt with.

Good shout. That case is indeed missing.
8846dd2954b43cb86a27dfeaf5bbb8c620e8184a

> That might require some extra SQL queries though — I haven't checked.

They do. That's the reason I didn't want to do it in this way.

If there's a particular control flow that reduces the number of introspective
queries even further, I'm in. However I don't want to penalise the most common
case (where all steps need to be performed) with one extra introspection query
per step.